libs/dyn/value.go

```diff
@@ -130,3 +139,90 @@ func (v Value) MarkAnchor() Value {
 func (v Value) IsAnchor() bool {
 	return v.anchor
 }
+
+func (v Value) IsScalarValueInString() bool {
+	if v.Kind() != KindString {
+		return false
+	}
+
+	// Parse value of the string and check if it's a scalar value.
+	// If it's a scalar value, we want to quote it.
+	switch v.MustString() {
+	case "true", "false":
+		return true
+	default:
+		_, err := parseNumber(v.MustString())
+		return err == nil
+	}
+}
+
+func (v Value) MarshalYAML() (interface{}, error) {
+	switch v.Kind() {
+	case KindMap:
+		m, _ := v.AsMap()
+		keys := maps.Keys(m)
+		// We're using location lines to define the order of keys in YAML.
+		// The location is set when we convert API response struct to config.Value representation
+		// See convert.convertMap for details
+		sort.SliceStable(keys, func(i, j int) bool {
+			return m[keys[i]].Location().Line < m[keys[j]].Location().Line
+		})
+
+		content := make([]*yaml.Node, 0)
+		for _, k := range keys {
+			item := m[k]
+			node := yaml.Node{Kind: yaml.ScalarNode, Value: k}
+			in, err := item.MarshalYAML()
+			c := in.(*yaml.Node)
+			if err != nil {
+				return nil, err
+			}
+			content = append(content, &node)
+			content = append(content, c)
+		}
+
+		return &yaml.Node{Kind: yaml.MappingNode, Content: content}, nil
+	case KindSequence:
+		s, _ := v.AsSequence()
+		content := make([]*yaml.Node, 0)
+		for _, item := range s {
+			in, err := item.MarshalYAML()
+			c := in.(*yaml.Node)
+			if err != nil {
+				return nil, err
+			}
+			content = append(content, c)
+		}
+		return &yaml.Node{Kind: yaml.SequenceNode, Content: content}, nil
+	case KindNil:
+		return &yaml.Node{Kind: yaml.ScalarNode, Value: "null"}, nil
+	case KindString:
+		// If the string is a scalar value (bool, int, float and etc.), we want to quote it.
+		if v.IsScalarValueInString() {
+			return &yaml.Node{Kind: yaml.ScalarNode, Value: v.MustString(), Style: yaml.DoubleQuotedStyle}, nil
+		}
+		return &yaml.Node{Kind: yaml.ScalarNode, Value: v.MustString()}, nil
+	case KindBool:
+		return &yaml.Node{Kind: yaml.ScalarNode, Value: fmt.Sprint(v.MustBool())}, nil
+	case KindInt:
+		return &yaml.Node{Kind: yaml.ScalarNode, Value: fmt.Sprint(v.MustInt())}, nil
+	case KindFloat:
+		return &yaml.Node{Kind: yaml.ScalarNode, Value: fmt.Sprint(v.MustFloat())}, nil
+	case KindTime:
+		return &yaml.Node{Kind: yaml.ScalarNode, Value: v.MustTime().UTC().String()}, nil
+	default:
+		// Panic because we only want to deal with known types.
+		panic(fmt.Sprintf("invalid kind: %d", v.Kind()))
+	}
+}
+
+func parseNumber(s string) (any, error) {
+	if i, err := strconv.ParseInt(s, 0, 64); err == nil {
+		return i, nil
+	}
+
+	if f, err := strconv.ParseFloat(s, 64); err == nil {
+		return f, nil
+	}
+	return nil, fmt.Errorf("invalid number: %s", s)
+}
```

Comment:
These new functions seem to belong in `yamlsaver` as well, given they're YAML specific.

Reply:
@pietern we need `MarshalYAML` to be `Value` method as we use yaml.encode to marshal it. Thus these functions here as well as otherwise we have a cycle dependency

Reply:
Understood.

We need this because the map passed to the `yamlsaver` function to save the file passes the incoming value directly to `yaml.Marshal`, so it only works if we implement `MarshalYAML` on the value type. But it does the same for `*yaml.Node`, so if we move this functionality into the `yamlsaver` package to turn a `dyn.Value` into a `yaml.Node`, we can keep all the serialization/deserialization logic outside of the core package.

Reply:
Done in the latest revision